pkg/applicationserver/applicationserver.go

```diff
@@ -690,27 +897,29 @@ func (as *ApplicationServer) handleDownlinkQueueInvalidated(ctx context.Context,
 }
 
 func (as *ApplicationServer) handleDownlinkNack(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, msg *ttnpb.ApplicationDownlink, link *link) error {
+	logger := log.FromContext(ctx)
 	client := ttnpb.NewAsNsClient(link.conn)
 	res, err := client.DownlinkQueueList(ctx, &ids, link.callOpts...)
 	if err != nil {
-		log.WithError(err).Warn("Failed to list the downlink queue for inserting nacked downlink message")
+		logger.WithError(err).Warn("Failed to list the downlink queue for inserting nacked downlink message")
 		registerDropDownlink(ctx, ids, msg, err)
 	} else {
-		queue := append([]*ttnpb.ApplicationDownlink{msg}, res.Downlinks...)
 		_, err := as.deviceRegistry.Set(ctx, ids,
 			[]string{
+				"pending_session",
 				"session",
 				"skip_payload_crypto",
 			},
 			func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
-				if err := as.recalculateDownlinkQueue(ctx, dev, nil, queue, msg.FCnt+1, link); err != nil {
+				queue := append([]*ttnpb.ApplicationDownlink{msg}, res.Downlinks...)
+				if err := as.recalculateDownlinkQueue(ctx, dev, link, dev.Session, queue, msg.FCnt+1); err != nil {
```

Comment:
Same as above, this neglects the downlinks for `dev.PendingSession` and will register them as dropped (as well as publish an event)

Reply:
Is this addressed?

Reply:
Yes, the `queue` is partitioned now before adding the `nacked` element to the front.